expo-updates breaks expo-filesystem on Android (API level 24+25)

### Minimal reproducible example

https://github.com/joshbuckley182/debug-copyasync

### What platform(s) does this occur on?

Android

### Where did you reproduce the issue?

in a development build

### Summary

When running on Android 7/7.1 (API level 24/25), the existence of the `expo-updates` package breaks `expo-file-system`. In particular, the `copyAsync` function breaks. 

In the MRE you'll see three commits:
1. Initial commit (https://github.com/joshbuckley182/debug-copyasync/commit/5aab27da60f6e8221b6a7f9c203c655829a3495f)
2. Add debug functions and `expo run:android` (https://github.com/joshbuckley182/debug-copyasync/commit/240dfa8107811052711431b4707bb5ddd371c788)
3. Add `expo-updates` (https://github.com/joshbuckley182/debug-copyasync/commit/face65fe9b49bfed97249e88412586eae6b9e4a1)

The MRE app contains 4 buttons which log to the console:
- "Check Document Directory" - logs the contents of `documentDirectory`
- "Check SQLite Directory" - logs the content of the SQLite directory
- "Remove SQLite Directory" - removes the SQLite directory
- "Copy Database" - Attempts to copy `test.db` asset into the SQLite directory

To reproduce:
- Check out commit `240dfa8` (second commit, working correctly)
- Run on an API level 24/25 device
- Click the buttons:
  - "Check Document Directory", and observe that no `SQLite` directory is logged
  - "Copy Database", and observe the log states the copy/from/to data, and ends with "Done"
  - "Check Document Directory", and observe that a `SQLite` directory now exists
  - "Check SQLite Directory", and observe that `test.db` exists
  - "Remove SQLite Directory", and observe that it logs "Done"
- Check out commit `face65f` (which adds `expo-updates`)
- Run on an API level 24/25 device
- Click the buttons:
  - "Check Document Directory", and observe that no `SQLite` directory is logged
  - "Copy Database", and observe the log states the copy/from/to data, **but doesn't log "Done"**
  - "Check Document Directory", and observe that there still isn't a `SQLite` directory


In my actual app I get this error which I don't see in the MRE, but the same "fix" of removing `expo-updates` resolves it.
```
 (NOBRIDGE) WARN  Possible Unhandled Promise Rejection (id: 0):
Error: Call to function 'ExponentFileSystem.copyAsync' has been rejected.
→ Caused by: java.lang.NoClassDefFoundError: Failed resolution of: [Ljava/nio/file/CopyOption;
```

### Environment

```text
npx expo-env-info

  expo-env-info 1.2.1 environment info:
    System:
      OS: macOS 15.1.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.9.0 - ~/.nvm/versions/node/v20.9.0/bin/node
      npm: 10.1.0 - ~/.nvm/versions/node/v20.9.0/bin/npm
    Managers:
      CocoaPods: 1.16.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.1, iOS 18.1, macOS 15.1, tvOS 18.1, visionOS 2.1, watchOS 11.1
    IDEs:
      Android Studio: 2024.2 AI-242.23339.11.2421.12700392
      Xcode: 16.1/16B40 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~52.0.17 => 52.0.17 
      react: 18.3.1 => 18.3.1 
      react-native: 0.76.3 => 0.76.3 
    npmGlobalPackages:
      eas-cli: 13.4.2
    Expo Workflow: managed
```


### Expo Doctor Diagnostics

```text
npx expo-doctor@latest
Enabled experimental React Native Directory checks. Unset the EXPO_DOCTOR_ENABLE_DIRECTORY_CHECK environment variable to disable this check.
✔ Check package.json for common issues
✔ Check Expo config for common issues
✔ Check native tooling versions
✔ Check if the project meets version requirements for submission to app stores
✔ Check for common project setup issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for app config fields that may not be synced in a non-CNG project
✔ Check npm/ yarn versions
✔ Check for issues with Metro config
✔ Validate packages against React Native Directory package metadata
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that packages match versions required by installed Expo SDK
✔ Check that native modules use compatible support package versions for installed Expo SDK

Didn't find any issues with the project!
```

--- a/packages/expo-file-system/CHANGELOG.md
+++ b/packages/expo-file-system/CHANGELOG.md
@@ -10,6 +10,8 @@
 
 ### 🐛 Bug fixes
 
+- Fix expo-updates breaking filesystem on Android API 24 and 25. ([#33694](https://github.com/expo/expo/pull/33694) by [@aleqsio](https://github.com/aleqsio))
+
 ### 💡 Others
 
 ## 18.0.6 - 2024-12-16

--- a/packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt
+++ b/packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt
@@ -291,9 +291,19 @@ open class FileSystemModule : Module() {
           val from = fromUri.toFile()
           val to = toUri.toFile()
           if (from.isDirectory) {
-            FileUtils.copyDirectory(from, to)
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+              FileUtils.copyDirectory(from, to)
+            } else {
+              // to be removed once Android SDK 25 support is dropped
+              from.copyRecursively(to, overwrite = true)
+            }
           } else {
-            FileUtils.copyFile(from, to)
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+              FileUtils.copyFile(from, to)
+            } else {
+              // to be removed once Android SDK 25 support is dropped
+              from.copyTo(to, overwrite = true)
+            }
           }
         }
 

